### PR TITLE
Add artifacts resolution on workspace projects

### PR DESCRIPTION
**Pull request by vrubezhny**
Add support for dependency (validation, completion) on workspace projects #77
    
Signed-off-by: Victor Rubezhny <vrubezhny@redhat.com>

### CODE REVIEW

File: lemminx-maven/pom.xml [hunk 1]
```diff
@@ -188,6 +188,16 @@
 				<enabled>true</enabled>
 			</releases>
 		</repository>
+		<repository>
```
Comment (mickaelistria) [inline lemminx-maven/pom.xml:191]: Please remove it as we use the release now.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java [hunk 2]
```diff
@@ -328,9 +330,16 @@ public MavenPluginManager getMavenPluginManager() {
...
+	
 	public Optional<RemoteRepositoryIndexSearcher> getIndexSearcher() {
 		initialize();
 		return Optional.ofNullable(indexSearcher);
 	}
+
+	public void didChangeWorkspaceFolders(URI[] added, URI[] removed) {
+		WorkspaceReader workspaceReader = mavenRequest.getWorkspaceReader();
+		if (workspaceReader instanceof MavenLemminxWorkspaceReader) {
+			// TODO
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java:342]: This should actually process all the workspaceFolders pom.xml files to add them in the cache.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+				.findAny();
+
+		if (matchingProject.isPresent()) {
+			MavenProject project = matchingProject.get();
+			File file = find(project, artifact);
+			if (file == null && project != project.getExecutionProject()) {
+				file = find(project.getExecutionProject(), artifact);
+			}
+
+//			System.out.println(getClass().getSimpleName() + ".findArtifact(" + artifact + "): "
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java:58]: Please remove.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+//					+ (file == null ? "null" : file.getAbsolutePath()));
+			return file;
+		}
+
+//		System.out.println(getClass().getSimpleName() + ".findArtifact(" + artifact + "): noll");
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java:63]: Please remove

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+ 
+	@Override
+	public List<String> findVersions(Artifact artifact) {
+		String key = ArtifactUtils.versionlessKey(artifact.getGroupId(), artifact.getArtifactId());
+		List<MavenProject> projects = plugin.getProjectCache().getProjects().stream()
+				.filter(p -> key.equals(ArtifactUtils.versionlessKey(p.getGroupId(), p.getArtifactId())))
+				.collect(Collectors.toList());
+
+		if (projects == null || projects.isEmpty()) {
+//			System.out.println(getClass().getSimpleName() + ".findVersions(" + artifact + "): []");
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java:75]: Please remove

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+		}
+
+		List<String> versions = new ArrayList<>();
+
+		for (MavenProject project : projects) {
+			if (find(project, artifact) != null) {
+				versions.add(project.getVersion());
+			}
+		}
+//		System.out.println(getClass().getSimpleName() + ".findVersions(" + artifact + "): " + versions.toString());
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java:86]: Please remove

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java [hunk 7]
```diff
@@ -624,64 +624,82 @@ private void internalCollectRemoteGAVCompletion(ICompletionRequest request, bool
...
-							}
-							return updatingItem;
-						case ARTIFACT_ID_ELT:
-							if (onlyPlugins) {
-								artifactInfosCollector
-										.addAll(indexSearcher.getPluginArtifacts(artifactToSearch, index));
-							} else {
+			if (indexSearcher!= null) {
+				try {
+					System.out.println("Step in into the lambda");
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:629]: Please remove.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java [hunk 8]
```diff
@@ -624,64 +624,82 @@ private void internalCollectRemoteGAVCompletion(ICompletionRequest request, bool
...
-										.map(groupId -> toCompletionItem(groupId, null, range))
-										.forEach(nonArtifactCollector::addCompletionItem);
-							}
-							return updatingItem;
-						case ARTIFACT_ID_ELT:
-							if (onlyPlugins) {
-								artifactInfosCollector
-										.addAll(indexSearcher.getPluginArtifacts(artifactToSearch, index));
-							} else {
+			if (indexSearcher!= null) {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:627]: What's actually changed here in the context of the dependency resolution? We should avoid unrelated changes to ease review and improve traceability.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/searcher/RemoteRepositoryIndexSearcher.java [hunk 9]
```diff
@@ -113,14 +113,19 @@ public RemoteRepositoryIndexSearcher(MavenLemminxExtension lemminxMavenPlugin, P
...
 			final IndexingContext context = initializeContext(repositoryUrl);
-			indexingContexts.put(repositoryUrl, context);
-			CompletableFuture<IndexingContext> future;
-			try {
-				future = updateIndex(context).thenApply(theVoid -> context);
-				indexDownloadJobs.put(context, future);
-				return future;
-			} catch (ComponentLookupException e) {
-				return CompletableFuture.failedFuture(e);
+			if (context != null) {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/searcher/RemoteRepositoryIndexSearcher.java:116]: What's actually changed here in the context of the dependency resolution? We should avoid unrelated changes to ease review and improve traceability.

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/SimpleModelTest.java [hunk 10]
```diff
@@ -284,6 +284,21 @@ public void testModuleDefinition() throws IOException, URISyntaxException {
 		assertTrue(definitionLinks.stream().anyMatch(link -> link.getTargetUri().equals(targetDocument.getDocumentURI())));
 	}
 
+	@Test
+	public void testModules() throws IOException, URISyntaxException {
+		List<Diagnostic> diagnosticsA = languageService.doDiagnostics(
+				createDOMDocument("/modules/module-a-pom.xml", languageService), 
+				new XMLValidationSettings(), () -> {});
+		diagnosticsA.stream().forEach(d -> System.out.println("    " + getClass().getSimpleName() + ":testModules(): Module A message: " + d.getMessage()));
```
Comment (mickaelistria) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/SimpleModelTest.java:292]: Please remove system.out

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/SimpleModelTest.java [hunk 11]
```diff
@@ -284,6 +284,21 @@ public void testModuleDefinition() throws IOException, URISyntaxException {
...
+		assertFalse(diagnosticsA.stream().anyMatch(diag -> (diag.getMessage().contains("ModuleA") || diag.getMessage().contains("ModuleB"))));
+
+		List<Diagnostic> diagnosticsB = languageService.doDiagnostics(
+				createDOMDocument("/modules/module-b-pom.xml", languageService), 
+				new XMLValidationSettings(), () -> {});
+		diagnosticsB.stream().forEach(d -> System.out.println("    " + getClass().getSimpleName() + ":testModules(): Module A message: " + d.getMessage()));
```
Comment (mickaelistria) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/SimpleModelTest.java:298]: please remove system.out

**Review by mickaelistria — CHANGES_REQUESTED** [review]

**Comment by mickaelistria** [comment]
Please also retitle the commit as the completion part doesn't seem to be in the scope of this commit; focus on "resolution".

File: lemminx-maven/pom.xml (hunk 1 above)
Comment (vrubezhny) [reply lemminx-maven/pom.xml:191]: ok

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java (hunk 3 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java:58]: ok

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java (hunk 4 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java:63]: ok

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java (hunk 5 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java:75]: ok

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java (hunk 7 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:629]: ok

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java (hunk 6 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxWorkspaceReader.java:86]: ok

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java (hunk 8 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:627]: The problem is that if the Central Maven repository us disabled a number of NPE is being reported during the tests which may make some tests to fail. So it was a try to prevent the appearing of those exceptions.

This code is not directly related to the issue and can be removed (or moved to another PR, if acceptable)

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/searcher/RemoteRepositoryIndexSearcher.java (hunk 9 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/searcher/RemoteRepositoryIndexSearcher.java:116]: Same as previous. Can be removed or moved to another PR, if acceptable

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/SimpleModelTest.java (hunk 10 above)
Comment (vrubezhny) [reply lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/SimpleModelTest.java:292]: ok

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/SimpleModelTest.java (hunk 11 above)
Comment (vrubezhny) [reply lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/SimpleModelTest.java:298]: ok

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java (hunk 8 above)
Comment (mickaelistria) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:627]: Having this in a separate change/PR would be great.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java (hunk 2 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java:342]: Does it mean that I have to recursively parse at least all the XML files that exist in the added folders in order to find and add pom files to the Project Cache?

Comment (mickaelistria) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java:342]: That would actually be nice. However, I think it's not necessary to implement this "TODO" for this commit, let's ignore it at the moment and we'll implement it in further commits.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java (hunk 8 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:627]: OK. I'll move it to a new commit

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java (hunk 2 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java:342]: ok

**Comment by vrubezhny** [comment]
@mickaelistria I've updated the PR according to your comments

**Comment by mickaelistria** [comment]
Perfect, thanks!
